homeassistant/components/asuswrt/config_flow.py

```diff
@@ -101,71 +127,98 @@ def _get_ip(host: str) -> str | None:
 
 
 class AsusWrtFlowHandler(ConfigFlow, domain=DOMAIN):
-    """Handle a config flow."""
+    """Handle a config flow for AsusWRT."""
 
     VERSION = 1
 
+    def __init__(self) -> None:
+        """Initialize the AsusWrt config flow."""
+        self._config_data: dict[str, Any] = {}
+        self._unique_id: str | None = None
+
     @callback
-    def _show_setup_form(
-        self,
-        user_input: dict[str, Any] | None = None,
-        errors: dict[str, str] | None = None,
-    ) -> FlowResult:
+    def _show_setup_form(self, error: str | None = None) -> FlowResult:
         """Show the setup form to the user."""
 
-        if user_input is None:
-            user_input = {}
+        user_input = self._config_data
 
-        adv_schema = {}
-        conf_password = vol.Required(CONF_PASSWORD)
         if self.show_advanced_options:
-            conf_password = vol.Optional(CONF_PASSWORD)
-            adv_schema[vol.Optional(CONF_PORT)] = cv.port
-            adv_schema[vol.Optional(CONF_SSH_KEY)] = str
+            add_schema = {
+                vol.Optional(CONF_PASSWORD): str,
+                vol.Optional(CONF_PORT): cv.port,
+                vol.Optional(CONF_SSH_KEY): str,
+            }
+        else:
+            add_schema = {vol.Required(CONF_PASSWORD): str}
 
         schema = {
             vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
             vol.Required(CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")): str,
-            conf_password: str,
-            vol.Required(CONF_PROTOCOL, default=PROTOCOL_SSH): vol.In(
-                {PROTOCOL_SSH: "SSH", PROTOCOL_TELNET: "Telnet"}
-            ),
-            **adv_schema,
-            vol.Required(CONF_MODE, default=MODE_ROUTER): vol.In(
-                {MODE_ROUTER: "Router", MODE_AP: "Access Point"}
+            **add_schema,
+            vol.Required(
+                CONF_PROTOCOL,
+                default=user_input.get(CONF_PROTOCOL, PROTOCOL_HTTPS),
+            ): SelectSelector(
+                SelectSelectorConfig(
+                    options=ALLOWED_PROTOCOL, translation_key="protocols"
+                )
             ),
         }
 
         return self.async_show_form(
             step_id="user",
             data_schema=vol.Schema(schema),
-            errors=errors or {},
+            errors={CONF_BASE: error} if error else None,
         )
 
     async def _async_check_connection(
         self, user_input: dict[str, Any]
     ) -> tuple[str, str | None]:
         """Attempt to connect the AsusWrt router."""
 
+        api: AsusWrtBridge
         host: str = user_input[CONF_HOST]
-        api = AsusWrtBridge.get_bridge(self.hass, user_input)
+        protocol = user_input[CONF_PROTOCOL]
+        error: str | None = None
+
+        conf = {**user_input, CONF_MODE: MODE_ROUTER}
```

Comment:
What does the CONF_MODE?

Reply:
CONF_MODE is an option required by legacy library that can be "Router" or "Access Point". Do not make any difference in connection, but use different logic at runtime. 
Because this is requested in a dedicated step when the user select SSH or TELNET protocol, during connection test I force this to "Router"